File: 2023/scala/src/main/scala/day03.scala
package day03

import locations.Directory.currentDir
import inputs.Input.loadFileSync
import scala.compiletime.ops.int

@main def part1: Unit =
  println(s"The solution is ${part1(loadInput())}")

@main def part2: Unit =
  println(s"The solution is ${part2(loadInput())}")

def loadInput(): String = loadFileSync(s"$currentDir/../../../../input/day03")

type Position = (Int, Int)

extension (p: Position)
  def row = p._1
  def col = p._2

// To store continuous digits
case class ContiguousDigits(digits: String, position: Position)

// To store continuous digits adjacent symbol. This is a part.
case class Part(digits: String, position: Position):
  val size = digits.size
  val number = digits.toInt

// An asterisk (*) with two adjacent parts and its location
case class Gear(position: Position, part1: Part, part2: Part):
  val ratio = part1.number * part2.number

val offsets =
  List((-1, -1), (-1, 0), (-1, 1), (0, -1), (0, 1), (1, -1), (1, 0), (1, 1))

case class EngineSchematic(input: String):

  val rows = input.linesIterator.map { _.toCharArray().toVector }.toVector

  val height = rows.size
  val width = rows(0).size
  val numbers =
    (for r <- 0 until rows.size yield {
      val (nums, (partCol, chars)) =
        rows(r).zipWithIndex.foldLeft(
          (List.newBuilder[ContiguousDigits], (-1, List[Char]()))
        ) { case ((nums, (partCol, chars)), (ch, col)) =>
          if ch.isDigit then
            (
              nums,
              if chars.isEmpty then (col, List(ch)) else (partCol, chars :+ ch)
            )
          else if chars.isEmpty then (nums, (-1, chars))
          else if chars.size > 0 then
            val part = ContiguousDigits(chars.mkString, (r, partCol))
            (nums += part, (-1, List()))
          else (nums, (-1, List()))
        }

      if chars.size > 0 then
        val part = ContiguousDigits(chars.mkString, (r, partCol))
        nums += part

      nums.result()
    }).flatten

  def inBounds(pos: Position) =
    pos.row >= 0 && pos.col >= 0 && pos.row < height && pos.col < width

  def isSymbol(pos: Position) =
    val (r, c) = pos
    rows(r)(c) != '.' && !rows(r)(c).isDigit

  def adjacentPositions(size: Int, position: Position) =

    val leftCellsOffset =
      List((-1, -1), (0, -1), (1, -1))
        .map((r, c) => (position.row + r, position.col + c))
     

    val topAndBottomCells =
      (0 until size)
        .flatMap(cc =>
          List(
            (position.row - 1, position.col + cc),
            (position.row + 1, position.col + cc)
          )
        )
     

    val rightCells =
      List((-1, 1), (0, 1), (1, 1))
        .map((r, c) => (position.row + r, position.col + c + size - 1))
        

    (leftCellsOffset ++ topAndBottomCells ++ rightCells).filter(inBounds).toList

  def symbolsAroundCells(position: Position, count: Int) =
    adjacentPositions(count, position)
      .filter(isSymbol)
      .map((r, c) => (rows(r)(c), (r, c)))

  val parts =
    numbers
      .filter(numInfo =>
        symbolsAroundCells(numInfo.position, numInfo.digits.size).nonEmpty
      )
      .map(numInfo => Part(numInfo.digits, numInfo.position))
      .toList

  val gears = parts
    .flatMap(part => {
      symbolsAroundCells(part.position, part.size)
        .find((ch, _) => ch == '*') match {
        case Some('*', asteriskPos) => Some(asteriskPos, part)
        case _                      => None
      }
    })
    .toSet
    .foldLeft(Map[Position, List[Part]]()) { case (m, (asteriskPos, part)) =>
      m.get(asteriskPos) match {
        case Some(nums) => m + (asteriskPos -> (part :: nums))
        case None       => m + (asteriskPos -> List(part))
      }
    }
    .flatMap {
      case (asteriskPos, List(part1, part2)) =>
        Some(Gear(asteriskPos, part1, part2))
      case _ => None
    }

def part1(input: String): String =

  val schematic = EngineSchematic(input)

  schematic.parts
    .map(_.number)
    .sum
    .toString()

def part2(input: String): String =

  val schematic = EngineSchematic(input)

  schematic.gears
    .map(_.ratio)
    .sum
    .toString()

@main def part12: Unit =

  val input = loadInput()
  val schematic = EngineSchematic(input)

  val sumOfPartNumbers = schematic.parts
    .map(_.number)
    .sum
    .toString()

  val sumOfGearRatio = schematic.gears
    .map(_.ratio)
    .sum
    .toString()

  println(s"Solution for part1 is $sumOfPartNumbers")
  println(s"Solution for part2 is $sumOfGearRatio")

File: 2023/haskell/src/Day03.hs
module Day03 (part1, part2) where

import Data.Char (isDigit, ord)
import Data.List (nub)
import qualified Data.Map as M

data Position = Position Int Int deriving (Show, Eq, Ord)

data Number = Number {number :: Int, numSize :: Int, numPosition :: Position} deriving (Show, Eq)

data Symbol = Symbol {symbol :: Char, symPosition :: Position} deriving (Show, Eq)

data Schematic = Schematic Int Int (M.Map Position Number) (M.Map Position Symbol) deriving (Show)

parseLine :: Int -> [(Char, Int)] -> M.Map Position Number -> M.Map Position Symbol -> (M.Map Position Number, M.Map Position Symbol)
parseLine _ [] nums symbols = (nums, symbols)
parseLine r (('.', _) : xs) nums symbols = parseLine r xs nums symbols
parseLine r xs@((x, c) : xt) nums symbols =
  if isDigit x
    then parseLine r rest (M.insert pos (Number {number = num, numSize = length digits, numPosition = Position r c}) nums) symbols
    else parseLine r xt nums (M.insert pos (Symbol {symbol = x, symPosition = Position r c}) symbols)
  where
    (digits, rest) = span (isDigit . fst) xs
    num = foldl (\t d -> t * 10 + ord d - 48) 0 (map fst digits)
    pos = Position r c

adjacentPositions :: Int -> Int -> Number -> [(Position, Number)]
adjacentPositions height width numInfo@Number {numPosition = Position r c, numSize = size} = do
  let left = filter inBounds [(r - 1, c - 1), (r, c - 1), (r + 1, c - 1)]
  let middle = concatMap (\s -> filter inBounds [(r - 1, c + s), (r + 1, c + s)]) [0 .. size - 1]
  let right = filter inBounds [(r - 1, c + size), (r, c + size), (r + 1, c + size)]
  map (\(rr, cc) -> (Position rr cc, numInfo)) $ left ++ middle ++ right
  where
    inBounds :: (Int, Int) -> Bool
    inBounds (rr, cc) = rr >= 0 && rr < height && cc >= 0 && cc < width

numbersNearSymbols :: Schematic -> [Number]
numbersNearSymbols (Schematic height width nums symbols) = do
  let numInfoList = M.map id nums
  let symbolPositions = M.keysSet symbols
  nub $ concatMap (map snd . filter (\(pos, _) -> pos `elem` symbolPositions) . adjacentPositions height width) numInfoList

numbersNearAsterisk :: Schematic -> M.Map Position [Number]
numbersNearAsterisk (Schematic height width nums symbols) = do
  let symbolPositions = M.keysSet $ M.filter ((== '*') . symbol) symbols
  let xs = concatMap (filter (\(pos, _) -> pos `elem` symbolPositions) . adjacentPositions height width) nums
  foldl (\m (pos, num) -> M.insertWith (++) pos [num] m) M.empty xs

createSchematic :: String -> Schematic
createSchematic input = do
  Schematic height width nums symbols
  where
    ls = lines input
    height = length ls
    width = length $ head ls
    (nums, symbols) =
      foldl (\(nums1, symbols1) (r, line) -> parseLine r (zip line [0 ..]) nums1 symbols1) (M.empty, M.empty) $ zip [0 ..] ls

part1 :: String -> IO ()
part1 input = do
  let sch = createSchematic input
  let numbers = numbersNearSymbols sch
  let sumOfIds = sum $ map number numbers
  print sumOfIds

part2 :: String -> IO ()
part2 input = do
  let sch = createSchematic input
  let numbers = numbersNearAsterisk sch
  let gears = M.filter ((== 2) . length) numbers
  let sumOfGearRatio = sum $ map ((\nums -> number (head nums) * number (nums !! 1)) . snd) $ M.assocs gears
  print sumOfGearRatio
